File: 프로그래머스/lv2/92341. 주차 요금 계산/주차 요금 계산.js
function solution(fees, records) {
  const infos = new Map();
  const totalTime = {};
  const [defaultTime, defaultFee, unitTime, unitFee] = fees;

  for (const record of records) {
    const [time, num, state] = record.split(" ");
    const [h, m] = time.split(":");
    const minute = Number(h) * 60 + Number(m);

    if (state === "IN") {
      infos.set(num, minute);
      continue;
    }

    const duration = minute - infos.get(num);
    totalTime[num] = totalTime[num] ? totalTime[num] + duration : duration;
    infos.delete(num);
  }

  // 출차 못한 차들 계산하기
  const last = 1439;
  for (const num of infos.keys()) {
    const duration = last - infos.get(num);
    totalTime[num] = totalTime[num] ? totalTime[num] + duration : duration;
  }

  // 정렬하기
  const sortedByCar = Object.keys(totalTime).sort((a, b) => Number(a) - Number(b));

  // 정답 출력
  const answer = [];
  for (const car of sortedByCar) {
    const totalFee = totalTime[car] > defaultTime ? 
          defaultFee + Math.ceil((totalTime[car] - defaultTime) / unitTime) * unitFee : defaultFee;
    answer.push(totalFee);
  }
  return answer;
}

File: 프로그래머스/lv2/92341. 주차 요금 계산/주차 요금 계산.py
import math

def solution(fees, records):
    [default, default_fee, per, per_fee] = fees
    total = {}
    current = {}
    end = 1439
    
    for record in records:
        time, car, state = tuple(record.split())
        h, m = tuple(map(int, time.split(':')))
        minutes = h * 60 + m
        
        if state == 'IN':
            if car not in total:
                total[car] = 0
            current[car] = minutes
        else:
            total[car] += minutes - current[car]
            current[car] = -1

    for key in current:
        if current[key] != -1:
            total[key] += end - current[key]
        
        if total[key] <= default:
            total[key] = default_fee
        else:
            total[key] = default_fee + math.ceil((total[key] - default) / per) * per_fee

    answer = [value for key, value in sorted(total.items(), key = lambda x: int(x[0]))]
    return answer